flow/src/test/java/com/sequenceiq/flow/helper/FlowPayloadConstructorChecker.java

```diff
@@ -0,0 +1,187 @@
+            checked.add(clazz);
+            checkForConstructor(clazz);
+        }
+    }
+
+    private void checkForConstructor(Class<?> clazz) {
+        Constructor<?>[] constructors = clazz.getDeclaredConstructors();
+        checkDefaultConstructor(clazz, constructors);
+        checkSoloBuilder(clazz, constructors);
+
+        Optional<Constructor<?>> maxParamCtor = getConstuctorWithMaximalParameterCount(constructors);
+        if (maxParamCtor.isPresent()) {
+            for (Class<?> paramType : maxParamCtor.get().getParameterTypes()) {
+                if (paramType.getPackageName().startsWith("com.sequenceiq.") &&
+                        !paramType.isEnum() &&
+                        !ClassUtils.isPrimitiveOrWrapper(paramType)) {
+
+                    checkForConstructorAndCache(paramType);
+                }
+            }
+        }
+    }
+
+    private void checkDefaultConstructor(Class<?> clazz, Constructor<?>[] constructors) {
+        if (hasDefaultConsturctor(constructors)) {
+            List<String> setterNames = checkAllSetters(clazz);
+            if (!setterNames.isEmpty()) {
+                validationErrors.add(String.format("Class %s with default constructor has setter(s) [%s] with no corresponding backing field(s)",
+                        clazz.getName(), String.join(",", setterNames)));
+            }
+
+            List<String> fieldNames = checkAllPrivateFields(clazz);
+            if (!fieldNames.isEmpty()) {
+                validationErrors.add(String.format("Class %s with default constructor has field(s) [%s] with no setter",
+                        clazz.getName(), String.join(",", fieldNames)));
+            }
+        }
+    }
+
+    private boolean hasDefaultConsturctor(Constructor<?>[] constructors) {
+        for (Constructor<?> c : constructors) {
+            if (c.getParameterCount() == 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void checkSoloBuilder(Class<?> clazz, Constructor<?>[] constructors) {
+        if (constructors.length == 1 &&
+                constructors[0].getParameterCount() == 1 &&
+                getBuilderClass(clazz).isPresent() &&
+                constructors[0].getParameterTypes()[0].equals(getBuilderClass(clazz).get())) {
+
+            validationErrors.add(
+                    String.format("Class %s has only a constructor for its builder thus it cannot be deserialized without proper annotation", clazz.getName()));
+        }
+    }
+
+    private Optional<Constructor<?>> getConstuctorWithMaximalParameterCount(Constructor<?>[] declaredConstructors) {
+        int maxParam = 0;
+        Constructor<?> maxParamCtor = null;
+        for (Constructor<?> c : declaredConstructors) {
+            if (c.getParameterCount() > maxParam) {
+                maxParam = c.getParameterCount();
+                maxParamCtor = c;
+            }
+        }
+        return Optional.ofNullable(maxParamCtor);
```

Comment:
Is it intentional that we are searching for constructors that have at least one arg? The current logic will never match the zero-arg constructor, hence if the target class only has the default zero-arg constructor and nothing else, `maxParamCtor` will indeed stay `null`.

If this is the desired goal, perhaps the method should be renamed by inserting `Positive` after `Maximal` (i.e. `getConstructorWithMaximalParameterCount`).

Reply:
renamed as there is a separate checked case for the default (zero-arg) ctor